text/0000-thread-safe-handles.md

```diff
@@ -0,0 +1,532 @@
+- Feature Name: Threadsafe Handles
+- Start Date: 2020-07-21
+- RFC PR: (leave this empty)
+- Neon Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Two new `Send` and `Sync` features are introduced: `EventQueue: Send + Sync` and `Root: Send`. These features work together to allow multi-threaded modules to interact with JavaScript.
+
+* `EventQueue`: Provides a `send` method for queueing a Rust closure to be executed by the JavaScript thread that created the `EventQueue`.
+* `Root<_>`: An external reference to JavaScript object that can be sent across threads. It may only be dereferenced by the JavaScript thread that created it.
+
+# Motivation
+[motivation]: #motivation
+
+High level event handlers have gone through [several](https://github.com/neon-bindings/rfcs/pull/25) [iterations](https://github.com/neon-bindings/rfcs/pull/28) without quite providing [ideal](https://github.com/neon-bindings/rfcs/issues/31) ergonomics or [safety](https://github.com/neon-bindings/neon/issues/551).
+
+The Threadsafe Handlers feature attempts to decompose the high level `EventHandler` API into lower level, more flexible primitives.
+
+These primitives can be used to build a higher level and _safe_ `EventHandler` API allowing experimentation with patterns outside of `neon`.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Neon provides a smart pointer type `Handle<'a, T: Value>` for referencing values on a JavaScript heap. These references are bound by the lifetime of `Context<'a>` to ensure they can only be used while the VM is locked in a synchronous neon function. These are neither `Send` or `Sync`.
+
+### `neon::handle::Root<T>`
```

Comment:
@dherman Do you think these belong in `neon::handle`? I think another option could be `neon::sync` to mirror `std::sync`.

Reply:
I think I'd lean towards `neon::handle` because they don't necessarily have to be used for multithreading use cases, right? Like, you can just use them to keep things alive for nonstandard lifetimes.

Reply:
Correct. And now that I think about it, calling them `Sync` is really just an implementation detail. Since they can only be read on the JavaScript main thread, they effectively _aren't_ `Sync`. I'll keep this one in `neon::handle`.

I still need a place to put `EventQueue`. Do you have thoughts on that? `sync` is an option, but it doesn't really align with `std::sync` because it's not a type wrapper.

Reply:
Maybe `neon::thread`?

Reply:
I like `neon::event` since it matches the name, and matches up with the JS concept of event handling and event queues.

Reply:
We ended up going with `neon::handle::Root` in the implementation, so I'll close this conversation.